core/vm/environment.go

```diff
@@ -17,110 +17,288 @@
+			caller.ReturnGas(gas)
+			return nil, nil
+		}
+
+		to = env.StateDB.CreateAccount(addr)
+	} else {
+		to = env.StateDB.GetAccount(addr)
+	}
+	env.Transfer(env.StateDB, caller.Address(), to.Address(), value)
+
+	// initialise a new contract and set the code that is to be used by the
+	// E The contract is a scoped environment for this execution context
+	// only.
+	contract := NewContract(caller, to, value, gas)
+	contract.SetCallCode(&addr, env.StateDB.GetCodeHash(addr), env.StateDB.GetCode(addr))
+	defer contract.Finalise()
+
+	ret, err := env.EVM().Run(contract, input)
+	// When an error was returned by the EVM or when setting the creation code
+	// above we revert to the snapshot and consume any gas remaining. Additionally
+	// when we're in homestead this also counts for code storage gas errors.
+	if err != nil {
+		contract.UseGas(contract.Gas)
+
+		env.StateDB.RevertToSnapshot(snapshotPreTransfer)
+	}
+	return ret, err
 }
 
-// Vm is the basic interface for an implementation of the EVM.
-type Vm interface {
-	// Run should execute the given contract with the input given in in
-	// and return the contract execution return bytes or an error if it
-	// failed.
-	Run(c *Contract, in []byte) ([]byte, error)
+// Take another's contract code and execute within our own context
+func (env *Environment) CallCode(caller ContractRef, addr common.Address, input []byte, gas, value *big.Int) ([]byte, error) {
+	if env.vmConfig.NoRecursion && env.Depth > 0 {
+		caller.ReturnGas(gas)
+
+		return nil, nil
+	}
+
+	// Depth check execution. Fail if we're trying to execute above the
+	// limit.
+	if env.Depth > int(params.CallCreateDepth.Int64()) {
+		caller.ReturnGas(gas)
+
+		return nil, DepthError
+	}
+	if !env.CanTransfer(env.StateDB, caller.Address(), value) {
+		caller.ReturnGas(gas)
+
+		return nil, fmt.Errorf("insufficient funds to transfer value. Req %v, has %v", value, env.StateDB.GetBalance(caller.Address()))
+	}
+
+	var (
+		snapshotPreTransfer = env.StateDB.Snapshot()
+		to                  = env.StateDB.GetAccount(caller.Address())
+	)
+	// initialise a new contract and set the code that is to be used by the
+	// E The contract is a scoped environment for this execution context
+	// only.
+	contract := NewContract(caller, to, value, gas)
+	contract.SetCallCode(&addr, env.StateDB.GetCodeHash(addr), env.StateDB.GetCode(addr))
+	defer contract.Finalise()
+
+	ret, err := env.EVM().Run(contract, input)
+	if err != nil {
+		contract.UseGas(contract.Gas)
+
+		env.StateDB.RevertToSnapshot(snapshotPreTransfer)
+	}
+
+	return ret, err
 }
 
-// Database is a EVM database for full state querying.
-type Database interface {
-	GetAccount(common.Address) Account
-	CreateAccount(common.Address) Account
+// Same as CallCode except sender and value is propagated from parent to child scope
+func (env *Environment) DelegateCall(caller ContractRef, addr common.Address, input []byte, gas *big.Int) ([]byte, error) {
+	if env.vmConfig.NoRecursion && env.Depth > 0 {
```

Comment:
Could you please add a comment that this is done for testing and doesn't execute during normal VM operation?